File: modules/folio-pdf-pages/ios/FolioPdfPagesModule.swift
import CoreGraphics
import ExpoModulesCore
import UIKit

struct FolioPdfPagesFailure: Error {
  let code: String
  let message: String
}

/// Renders one JPEG thumbnail per PDF page with Core Graphics. Pages are
/// rasterized strictly one at a time inside their own autorelease pool, so a
/// long scan costs the memory of a single page rather than of the whole
/// document.
public final class FolioPdfPagesModule: Module {
  private static let minimumWidth = 32
  private static let maximumWidth = 2_048
  /// Caps the long edge so a pathological page aspect ratio cannot grow the
  /// single bitmap beyond `maxWidth * 2048 * 4` bytes.
  private static let maximumHeight: CGFloat = 2_048
  private static let maximumPages = 10_000
  private static let jpegQuality: CGFloat = 0.8

  public func definition() -> ModuleDefinition {
    Name("FolioPdfPages")

    AsyncFunction("renderPageThumbnails") {
      (pdfUri: String, outputDirectoryUri: String, maxWidth: Int, promise: Promise) in
      do {
        let rendered = try Self.renderThumbnails(
          pdfUri: pdfUri,
          outputDirectoryUri: outputDirectoryUri,
          maxWidth: maxWidth
        )
        promise.resolve(rendered)
      } catch {
        Self.reject(promise, error)
      }
    }
  }

  private static func renderThumbnails(
    pdfUri: String,
    outputDirectoryUri: String,
    maxWidth: Int
  ) throws -> [String: Any] {
    guard maxWidth >= minimumWidth, maxWidth <= maximumWidth else {
      throw FolioPdfPagesFailure(
        code: "WIDTH_UNSUPPORTED",
        message: "The requested thumbnail width is outside the supported range."
      )
    }

    let sourceUrl = try fileUrl(pdfUri)
    guard FileManager.default.fileExists(atPath: sourceUrl.path) else {
      throw FolioPdfPagesFailure(
        code: "SOURCE_UNAVAILABLE",
        message: "The PDF is no longer available on this device."
      )
    }

    let outputUrl = try fileUrl(outputDirectoryUri)
    do {
      try FileManager.default.createDirectory(at: outputUrl, withIntermediateDirectories: true)
    } catch {
      throw FolioPdfPagesFailure(
        code: "OUTPUT_UNAVAILABLE",
        message: "The page preview directory could not be created."
      )
    }

    guard let document = CGPDFDocument(sourceUrl as CFURL) else {
      throw FolioPdfPagesFailure(
        code: "PDF_MALFORMED",
        message: "This PDF could not be read for page previews."
      )
    }
    guard !document.isEncrypted || document.isUnlocked else {
      throw FolioPdfPagesFailure(
        code: "PDF_PASSWORD_PROTECTED",
        message: "This PDF is password protected, so its pages cannot be previewed."
      )
    }

    let pageCount = document.numberOfPages
    guard pageCount >= 1 else {
      throw FolioPdfPagesFailure(
        code: "PDF_MALFORMED",
        message: "This PDF has no renderable pages."
      )
    }
    guard pageCount <= maximumPages else {
      throw FolioPdfPagesFailure(
        code: "PDF_TOO_MANY_PAGES",
        message: "This PDF has more pages than the editor can preview."
      )
    }

    var pages: [[String: Any]] = []
    pages.reserveCapacity(pageCount)
    for number in 1...pageCount {
      let rendered = try autoreleasepool {
        try renderPage(
          document: document,
          number: number,
          outputDirectory: outputUrl,
          maxWidth: maxWidth
        )
      }
      pages.append(rendered)
    }
    return ["pageCount": pageCount, "pages": pages]
  }

  private static func renderPage(
    document: CGPDFDocument,
    number: Int,
    outputDirectory: URL,
    maxWidth: Int
  ) throws -> [String: Any] {
    guard let page = document.page(at: number) else {
      throw FolioPdfPagesFailure(
        code: "PDF_MALFORMED",
        message: "A page in this PDF could not be opened for preview."
      )
    }

    // A page without a crop box falls back to its media box; both are given in
    // the PDF coordinate space, before the page's own /Rotate is applied.
    let cropBox = page.getBoxRect(.cropBox)
    let boxType: CGPDFBox = cropBox.isEmpty ? .mediaBox : .cropBox
    let box = cropBox.isEmpty ? page.getBoxRect(.mediaBox) : cropBox
    guard box.width > 0, box.height > 0 else {
      throw FolioPdfPagesFailure(
        code: "PDF_MALFORMED",
        message: "A page in this PDF has no usable size."
      )
    }

    let quarterTurn = abs(page.rotationAngle) % 180 == 90
    let uprightWidth = quarterTurn ? box.height : box.width
    let uprightHeight = quarterTurn ? box.width : box.height

    var width = CGFloat(maxWidth)
    var height = max(1, (uprightHeight * width / uprightWidth).rounded())
    if height > maximumHeight {
      height = maximumHeight
      width = max(1, (uprightWidth * height / uprightHeight).rounded())
    }
    let pixelSize = CGSize(width: width, height: height)
    let target = CGRect(origin: .zero, size: pixelSize)

    let format = UIGraphicsImageRendererFormat.default()
    // One image pixel per point: the caller already asked for a pixel width.
    format.scale = 1
    format.opaque = true
    let image = UIGraphicsImageRenderer(size: pixelSize, format: format).image { context in
      let cgContext = context.cgContext
      cgContext.setFillColor(UIColor.white.cgColor)
      cgContext.fill(target)
      // Core Graphics draws PDF pages bottom-up; flip into the UIKit frame
      // before applying the page's own box and rotation transform.
      cgContext.translateBy(x: 0, y: pixelSize.height)
      cgContext.scaleBy(x: 1, y: -1)
      cgContext.concatenate(page.getDrawingTransform(
        boxType,
        rect: target,
        rotate: 0,
        preserveAspectRatio: true
      ))
      cgContext.drawPDFPage(page)
    }

    guard let data = image.jpegData(compressionQuality: jpegQuality) else {
      throw FolioPdfPagesFailure(
        code: "OUTPUT_UNAVAILABLE",
        message: "A page preview could not be encoded."
      )
    }
    let destination = outputDirectory.appendingPathComponent("page-\(number).jpg")
    do {
      try data.write(to: destination, options: .atomic)
    } catch {
      throw FolioPdfPagesFailure(
        code: "OUTPUT_UNAVAILABLE",
        message: "A page preview could not be written to the private cache."
      )
    }

    return [
      "page": number,
      "uri": destination.absoluteString,
      "width": Int(pixelSize.width),
      "height": Int(pixelSize.height),
    ]
  }

  /// Accepts only `file://` URLs inside the application container, so the
  /// renderer can never be pointed at another application's files.
  private static func fileUrl(_ value: String) throws -> URL {
    guard let url = URL(string: value), url.isFileURL else {
      throw FolioPdfPagesFailure(
        code: "PATH_NOT_PRIVATE",
        message: "Only app-private file URLs are supported."
      )
    }
    let resolved = url.standardizedFileURL.resolvingSymlinksInPath()
    let container = URL(fileURLWithPath: NSHomeDirectory())
      .standardizedFileURL
      .resolvingSymlinksInPath()
    guard resolved.path == container.path
      || resolved.path.hasPrefix(container.path + "/") else {
      throw FolioPdfPagesFailure(
        code: "PATH_NOT_PRIVATE",
        message: "The file URL is outside Folio's private storage."
      )
    }
    return resolved
  }

  private static func reject(_ promise: Promise, _ error: Error) {
    if let failure = error as? FolioPdfPagesFailure {
      promise.reject("ERR_FOLIO_PDF_PAGES_\(failure.code)", failure.message)
      return
    }
    promise.reject(
      "ERR_FOLIO_PDF_PAGES_RENDER_FAILED",
      "The PDF pages could not be rendered on this device."
    )
  }
}

File: modules/folio-pdf-pages/android/src/main/java/app/folio/pdfpages/FolioPdfPagesModule.kt
package app.folio.pdfpages

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.ParcelFileDescriptor
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.IOException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.math.max
import kotlin.math.roundToInt

private const val MINIMUM_WIDTH = 32
private const val MAXIMUM_WIDTH = 2_048
// Caps the long edge so a pathological page aspect ratio cannot grow the single
// bitmap beyond `maxWidth * 2048 * 4` bytes (about 3 MiB at the editor's width).
private const val MAXIMUM_HEIGHT = 2_048
private const val MAXIMUM_PAGES = 10_000
private const val JPEG_QUALITY = 80

private class FolioPdfPagesFailure(val failureCode: String, message: String) : Exception(message)

/**
 * Renders one JPEG thumbnail per PDF page with the platform renderer. Pages are
 * rasterized strictly one at a time: a single ARGB_8888 bitmap exists at any
 * moment and is recycled before the next page is opened, so a long scan costs
 * the memory of one page instead of the memory of the whole document.
 */
class FolioPdfPagesModule : Module() {
  // A single worker keeps concurrent editor sessions from rasterizing at the
  // same time, which would defeat the one-bitmap-at-a-time guarantee.
  private val executor: ExecutorService = Executors.newSingleThreadExecutor()

  private val context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("FolioPdfPages")

    AsyncFunction("renderPageThumbnails") {
      pdfUri: String,
      outputDirectoryUri: String,
      maxWidth: Int,
      promise: Promise,
      ->
      executor.execute {
        try {
          promise.resolve(renderThumbnails(pdfUri, outputDirectoryUri, maxWidth))
        } catch (error: Throwable) {
          reject(promise, error)
        }
      }
    }

    OnDestroy {
      executor.shutdown()
    }
  }

  private fun renderThumbnails(
    pdfUri: String,
    outputDirectoryUri: String,
    maxWidth: Int,
  ): Map<String, Any> {
    if (maxWidth < MINIMUM_WIDTH || maxWidth > MAXIMUM_WIDTH) {
      throw FolioPdfPagesFailure(
        "WIDTH_UNSUPPORTED",
        "The requested thumbnail width is outside the supported range.",
      )
    }

    val source = privateFile(pdfUri)
    if (!source.isFile || source.length() <= 0L) {
      throw FolioPdfPagesFailure(
        "SOURCE_UNAVAILABLE",
        "The PDF is no longer available on this device.",
      )
    }

    val outputDirectory = privateFile(outputDirectoryUri)
    if (!outputDirectory.exists() && !outputDirectory.mkdirs()) {
      throw FolioPdfPagesFailure(
        "OUTPUT_UNAVAILABLE",
        "The page preview directory could not be created.",
      )
    }
    if (!outputDirectory.isDirectory) {
      throw FolioPdfPagesFailure(
        "OUTPUT_UNAVAILABLE",
        "The page preview destination is not a directory.",
      )
    }

    val descriptor = try {
      ParcelFileDescriptor.open(source, ParcelFileDescriptor.MODE_READ_ONLY)
    } catch (error: IOException) {
      throw FolioPdfPagesFailure(
        "SOURCE_UNREADABLE",
        "The PDF could not be opened for page previews.",
      )
    }

    var renderer: PdfRenderer? = null
    try {
      renderer = try {
        PdfRenderer(descriptor)
      } catch (error: SecurityException) {
        throw FolioPdfPagesFailure(
          "PDF_PASSWORD_PROTECTED",
          "This PDF is password protected, so its pages cannot be previewed.",
        )
      } catch (error: IOException) {
        throw FolioPdfPagesFailure(
          "PDF_MALFORMED",
          "This PDF could not be read for page previews.",
        )
      }

      val pageCount = renderer.pageCount
      if (pageCount < 1) {
        throw FolioPdfPagesFailure("PDF_MALFORMED", "This PDF has no renderable pages.")
      }
      if (pageCount > MAXIMUM_PAGES) {
        throw FolioPdfPagesFailure(
          "PDF_TOO_MANY_PAGES",
          "This PDF has more pages than the editor can preview.",
        )
      }

      val pages = ArrayList<Map<String, Any>>(pageCount)
      for (index in 0 until pageCount) {
        pages.add(renderPage(renderer, index, outputDirectory, maxWidth))
      }
      return mapOf("pageCount" to pageCount, "pages" to pages)
    } finally {
      try {
        renderer?.close()
      } catch (ignored: Throwable) {
        // A close failure must not mask the rendering outcome.
      }
      try {
        // PdfRenderer.close() already released this descriptor on the happy
        // path; closing an already-closed descriptor is a no-op.
        descriptor.close()
      } catch (ignored: IOException) {
        // Ignored for the same reason.
      }
    }
  }

  private fun renderPage(
    renderer: PdfRenderer,
    index: Int,
    outputDirectory: File,
    maxWidth: Int,
  ): Map<String, Any> {
    var page: PdfRenderer.Page? = null
    var bitmap: Bitmap? = null
    try {
      val current = try {
        renderer.openPage(index)
      } catch (error: Throwable) {
        throw FolioPdfPagesFailure(
          "PDF_MALFORMED",
          "A page in this PDF could not be opened for preview.",
        )
      }
      page = current

      val sourceWidth = current.width
      val sourceHeight = current.height
      if (sourceWidth <= 0 || sourceHeight <= 0) {
        throw FolioPdfPagesFailure(
          "PDF_MALFORMED",
          "A page in this PDF has no usable size.",
        )
      }

      var width = maxWidth
      var height = max(
        1,
        (sourceHeight.toDouble() * width.toDouble() / sourceWidth.toDouble()).roundToInt(),
      )
      if (height > MAXIMUM_HEIGHT) {
        height = MAXIMUM_HEIGHT
        width = max(
          1,
          (sourceWidth.toDouble() * height.toDouble() / sourceHeight.toDouble()).roundToInt(),
        )
      }

      val pageBitmap = try {
        Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
      } catch (error: OutOfMemoryError) {
        throw FolioPdfPagesFailure(
          "RENDER_OUT_OF_MEMORY",
          "This device could not allocate a preview for one of the pages.",
        )
      }
      bitmap = pageBitmap
      // PdfRenderer composites onto the destination, so an opaque white page
      // needs an explicitly white bitmap rather than the default transparency.
      pageBitmap.eraseColor(Color.WHITE)
      current.render(pageBitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
      current.close()
      page = null

      val destination = File(outputDirectory, "page-${index + 1}.jpg")
      val compressed = destination.outputStream().use { stream ->
        val ok = pageBitmap.compress(Bitmap.CompressFormat.JPEG, JPEG_QUALITY, stream)
        stream.flush()
        ok
      }
      if (!compressed) {
        throw FolioPdfPagesFailure(
          "OUTPUT_UNAVAILABLE",
          "A page preview could not be written to the private cache.",
        )
      }

      return mapOf(
        "page" to index + 1,
        "uri" to Uri.fromFile(destination).toString(),
        "width" to width,
        "height" to height,
      )
    } finally {
      try {
        page?.close()
      } catch (ignored: Throwable) {
        // Ignored: the renderer is closed by the caller either way.
      }
      bitmap?.recycle()
    }
  }

  /**
   * Accepts only `file://` URLs inside Folio's own sandbox, so the renderer can
   * never be pointed at shared storage or another application's files.
   */
  private fun privateFile(value: String): File {
    val uri = Uri.parse(value)
    val path = uri.path
    if (uri.scheme != "file" || path.isNullOrEmpty()) {
      throw FolioPdfPagesFailure(
        "PATH_NOT_PRIVATE",
        "Only app-private file URLs are supported.",
      )
    }
    val file = File(path).canonicalFile
    val roots = listOf(context.filesDir, context.cacheDir, context.noBackupFilesDir)
      .map { it.canonicalFile }
    if (!roots.any { file.path == it.path || file.path.startsWith(it.path + File.separator) }) {
      throw FolioPdfPagesFailure(
        "PATH_NOT_PRIVATE",
        "The file URL is outside Folio's private storage.",
      )
    }
    return file
  }

  private fun reject(promise: Promise, error: Throwable) {
    val failure = error as? FolioPdfPagesFailure
    promise.reject(
      "ERR_FOLIO_PDF_PAGES_${failure?.failureCode ?: "RENDER_FAILED"}",
      failure?.message ?: "The PDF pages could not be rendered on this device.",
      null,
    )
  }
}
